### PR TITLE
i#4504: Implement instr_invert_predicate for AArch64

### PR DESCRIPTION
**Pull request by yury-khrustalev**
Add implementation of `instr_invert_predicate` for AArch64.
This change is required to complete #4500 to allow creation of conditional instructions.

Fixes: #4504

### CODE REVIEW

**Comment by yury-khrustalev** [comment]
VS builds failed with "The remote server returned an error: (404) Not Found." error, could you please restart the builds, thank you!

**Comment by derekbruening** [comment]
> VS builds failed with "The remote server returned an error: (404) Not Found." error, could you please restart the builds, thank you!

Please see https://docs.github.com/en/free-pro-team@latest/actions/managing-workflow-runs/re-running-a-workflow

**Comment by yury-khrustalev** [comment]
> Please see https://docs.github.com/en/free-pro-team@latest/actions/managing-workflow-runs/re-running-a-workflow

@derekbruening, thank you, I know how to use GitHub actions, but I don't have "Re-run all jobs" button in this repository (probably due to not having permissions or something along these lines).

**Comment by yury-khrustalev** [comment]
Also, it looks like there is a problem with URL used to download Doxygen for windows builds.

**Comment by derekbruening** [comment]
> Also, it looks like there is a problem with URL used to download Doxygen for windows builds.

If the doxygen URL changed, then a re-run won't fix it.  If you have the new URL, submit a separate PR to fix it?

**Comment by yury-khrustalev** [comment]
The URL which is no longer available is

    http://doxygen.nl/files/doxygen-1.8.19.windows.x64.bin.zip

Probably version 1.8.19 is considered too old. The next version 1.8.20 is [available](https://doxygen.nl/files/doxygen-1.8.20.windows.x64.bin.zip), though. Or the most recent version 1.9.0 can be used instead (I'm not sure why 1.8.19 version is currently used in particular):

    https://doxygen.nl/files/doxygen-1.9.0.windows.x64.bin.zip

>  If you have the new URL, submit a separate PR to fix it?

If I submitted a new PR, the builds would obviously fail and the PR wouldn't be able to merge. I think it's easier for a maintainer to fix the URL directly.

**Comment by yury-khrustalev** [comment]
>  If you have the new URL, submit a separate PR to fix it?

Was there a reason why version 1.8.19 is used? Would it not work with 1.9.0? I simply don't have appropriate devices to test.

**Comment by derekbruening** [comment]
> > If you have the new URL, submit a separate PR to fix it?
> 
> Was there a reason why version 1.8.19 is used? Would it not work with 1.9.0?

I recall using chocolatey to download the latest w/o having to specify its version but chocolately was flaky: looks like #4000.  So we switched to a hardcoded version to solve the choco problems and probably just put in the latest version at the time, which then suffers from breaking when the link disappears.  Suggestions for a better way are welcome.

> I simply don't have appropriate devices to test.

I think anyone would test it the same way as you: make a PR and see if Windows turns green on the PR.  Or wait and someone else will hit it and have to fix it to make progress.

**Comment by yury-khrustalev** [comment]
@derekbruening, could you please review this when you have time. This comes from the discussion in #4500 and is required to rework that change according to what we decided. Thank you.

File: core/ir/instr.h [hunk 1]
```diff
@@ -1351,17 +1351,22 @@ DR_API
...
 
+#ifdef AARCHXX
 DR_API
 /**
  * Returns the DR_PRED_ constant that represents the opposite condition
  * from \p pred.  A valid conditional branch predicate must be passed (i.e.,
- * not #DR_PRED_NONE, DR_PRED_AL, or DR_PRED_OP).
- * \note ARM-only.
+ * not #DR_PRED_NONE, DR_PRED_AL, or DR_PRED_OP for ARM and not #DR_PRED_NONE,
+ * DR_PRED_AL, or DR_PRED_NV for AArch64).
```
Comment (derekbruening) [inline core/ir/instr.h:1362]: nit: missing doxygen links for these two (the `#`)
Looks like the original ARM is also missing them for 2 entries: if you could add those too.

**Review by derekbruening — APPROVED** [review]
(I see there's no existing test for the ARM version but it could be nice to add some sanity check to ir_aarch64.c or sthg.)

**Comment by derekbruening** [comment]
> Fixes: #4504 by adding implementation of instr_invert_predicate for AArch64.

The Fixes has to be on a line by itself.  See https://github.com/DynamoRIO/dynamorio/wiki/Code-Reviews#commit-messages where we always put it at the bottom.

File: core/ir/instr.h (hunk 1 above)
Comment (yury-khrustalev) [reply core/ir/instr.h]: This is fixed now.

**Pull request merged**